src/provider.c

```diff
@@ -300,9 +300,9 @@ static int ps_prov_init(const OSSL_CORE_HANDLE *handle,
 	}
 	ps_pctx_debug(pctx, "pctx: %p, forward: %s", pctx, pctx->fwd.name);
 
-	if (pkcs11_module_init(&pctx->pkcs11, module, module_args, &pctx->dbg) != OSSL_RV_OK) {
+	if (pkcs11_module_load(&pctx->pkcs11, module, module_args, &pctx->dbg) != OSSL_RV_OK) {
```

Comment:
In general it is a very good thing to delay the PKCS#11 initialization, but unfortunately it does not really help the way it is now.

`ps_prov_init()` calls `pkcs11_module_load()` which calls `module_info()` which calls `module_ensure()` which still initializes PKCS#11 in the provider init routine. So instead of a implicit PKCS#11 init you now have an implicit PKCS#11 init within the provider init function :-) 

Please check if you really need to call any PKCS#11 function during provider init, or if you can delay that also.

Reply:
IMHO, there are 2 options. a) remove `module_info()` completely or b) call it directly after `C_Initialize()`.
I personally would slightly prefer b).

Maybe we can also return early in module_info(), if the level is < dbg. It does not make any sense to call C_GetInfo(), if the result is not used.

Reply:
So module_info is just for debugging? 
But yes, I would also prefer option b). 
And sure, you could even skip it if debugging is not enabled.

Reply:
I will change this in the next version.

Reply:
> Maybe we can also return early in module_info(), if the level is < dbg. It does not make any sense to call C_GetInfo(), if the result is not used.

correction: `module_info()` uses the INFO level, so the early exit will be on `< INFO`, not on `< DEBUG` level.